com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java

```diff
@@ -104,16 +107,31 @@ public AdvancedViatraQueryEngine getInitializedEngineChecked() {
 	}
 	
 	private Optional<AdvancedViatraQueryEngine> initializeEngine() {
-		if(!isInitialized && ProjectUtilities.isLoaded(project.getPrimaryProject())) {
-			engine = Optional.of(createQueryEngine(project, notifiers));
-			notifiers = new Notifier[0];
-			isInitialized = true;
-		}
 		if(!isInitialized) {
-			LOGGER.warn(MESSAGE_ENGINE_NOT_READY);
-		} else {
-			initializationActions.forEach(action -> action.accept(engine.get()));
-			initializationActions.clear();
+			synchronized(this) {
+				if(!isInitialized && ProjectUtilities.isLoaded(project.getPrimaryProject())) {
+					engine = Optional.of(createQueryEngine(project, notifiers));
+				}
+				isInitialized = engine.map(e -> {
+						boolean thereWasException = false;
+						try {
+							e.getBaseIndex().coalesceTraversals(() -> {
+								initializationActions.forEach(action -> action.accept(e));
+								return null;
+							});
+							initializationActions.clear();
+							notifiers = new Notifier[0];
+						} catch (InvocationTargetException ite) {
+							LOGGER.warn(MESSAGE_ENGINE_PREPARE_ACTION_ERROR);
+							engine = null;
+							thereWasException = true;
+						}
+						return !thereWasException;
```

Comment:
The control flow in this method is really strange (it might be correct, but it is strange). In case an exception was thrown during initialization, we state that the engine is not initialized throw away the engine created, but keep all the previously saved actions for a second try (do we hope that the exception will get resolved at that point?)...

Question: is this behavior consistent with what happens when the engine initialization throws a ViatraQueryException (of course a different error message is fine, but we should handle these errors the same way).

Reply:
I think it is possible that we cannot initialize the engine correctly at some point but we can later (based on what I have seen from MagicDraw).
And yes, it is consistent: we throw a ViatraQueryException if the VIATRA engine initialization is not successful which is true in two cases: a) if it cannot be created or b) the execution of an action fails.